stan/math/prim/fun/value_of.hpp

```diff
@@ -10,127 +10,56 @@ namespace stan {
 namespace math {
 
 /**
- * Return the value of the specified scalar argument
- * converted to a double value.
+ * Inputs that are arithmetic types or containers of airthmetric types
+ * are returned from value_of unchanged
  *
- * <p>See the <code>primitive_value</code> function to
- * extract values without casting to <code>double</code>.
- *
- * <p>This function is meant to cover the primitive types. For
- * types requiring pass-by-reference, this template function
- * should be specialized.
- *
- * @tparam T type of scalar.
- * @param x scalar to convert to double
- * @return value of scalar cast to double
- */
-template <typename T, require_arithmetic_t<T>* = nullptr>
-inline double value_of(const T x) {
-  return static_cast<double>(x);
-}
-
-/**
- * Return the specified argument.
- *
- * <p>See <code>value_of(T)</code> for a polymorphic
- * implementation using static casts.
- *
- * <p>This inline pass-through no-op should be compiled away.
- *
- * @param x value
- * @return input value
- */
-template <>
-inline double value_of<double>(double x) {
-  return x;
+ * @tparam T Input type
+ * @param[in] x Input argument
+ * @return Forwarded input argument
+ **/
+template <typename T,
+	  require_arithmetic_t<stan::return_type_t<T>>* = nullptr>
```

Comment:
```suggestion
	  require_st_arithmetic_t<T>* = nullptr>
```

Reply:
Question on this. The reason I did `return_type` was cause I know `return_type` will burrow down through all the template types until it hits a scalar-like thing. Will the `scalar_type` type traits do this same thing?

And why not `require_vt_arithmetic_t`?

Reply:
(And by burrow I mean I know if I have `std::vector<std::vector<std::vector<std::vector<double>>>>` I'll get `double` back (or however many layers of templated stuff)

Reply:
Well, the reason `return_type` does that is because it uses `scalar_type` underneath. So it will work just as well. 

On the other hand vt in `require_vt_arithmetic_t` stands for `value_type`, which would go just one level down.